client/hwlib/src/collectors/cpuid.rs

```diff
@@ -0,0 +1,115 @@
+/* Copyright 2024 Canonical Ltd.
+ * All rights reserved.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+use anyhow::Result;
+use smbioslib;
+
+pub(super) fn get_cpuid(proc_info: &smbioslib::SMBiosProcessorInformation) -> Result<String> {
+    let cpu_id = proc_info
+        .processor_id()
+        .map(|id| format!("{:x?}", id))
+        .ok_or("Processor ID not found")
+        .unwrap();
+    Ok(cpu_id)
+}
+
+pub(super) fn convert_cpu_codename(cpu_id: &str) -> Result<String> {
+    if cpu_id.is_empty() {
+        return Ok(String::from("Unknown"));
```

Comment:
It's a bit strange to have "Unknown" be a positive outcome (`Ok`) rather than a negative one (`Err(...)`)

Reply:
On the C3 side, we return empty string if CPU ID is empty: https://github.com/canonical/hexr/blob/9084e69e88726158f5bd4e1947b3a3916afd33c0/apps/hardware/models.py#L1896-L1897

I've updated it to return an empty string as well, but it's still the `Ok` state. So it's fine to send the request with an empty string if we cannot find a CPU ID. Do you think we should change it to `Err` state?

(The data from C3 is used to match the data that we collect from the client. The hwapi server side just creates the replica of C3 DB, so we need the CPU codename data to be parsed the same way C3 parses it)

Reply:
If you know that these functions don't have to make intuitive sense outside of the context of outputting the data exactly like C3 then it's fine. If this is meant for a potentially wider use case then it could be worth returning an `Err` here and transforming that error into an empty string in some other piece of (C3-specific) logic.

Another thing to consider is that if you have a function that always returns `Ok(...)` no matter which path is taken, then you might as well not return a `Result`  but just return the thing wrapped inside.